marlowe-playground-client/src/Component/MetadataTab/State.purs

```diff
@@ -1,49 +1,69 @@
+import Record (merge) as Record
+import Type.Prelude (Proxy(..))
 
 carryMetadataAction ::
   forall m.
   MonadAff m =>
   MonadAsk Env m =>
   MetadataAction ->
   HalogenM State Action ChildSlots Void m Unit
-carryMetadataAction action = do
-  modifying (_contractMetadata) case action of
-    SetContractName name -> set _contractName name
-    SetContractType typeName -> set _contractType typeName
-    SetContractShortDescription description -> set _contractShortDescription description
-    SetContractLongDescription description -> set _contractLongDescription description
-    SetRoleDescription tokenName description -> over _roleDescriptions $ Map.insert tokenName description
-    DeleteRoleDescription tokenName -> over _roleDescriptions $ Map.delete tokenName
-    SetSlotParameterDescription slotParam description -> over _slotParameterDescriptions $ OMap.insert slotParam description
-    DeleteSlotParameterDescription slotParam -> over _slotParameterDescriptions $ OMap.delete slotParam
-    SetValueParameterDescription valueParameterName description -> over _valueParameterInfo $ updateValueParameterInfo (setValueParameterDescription description) valueParameterName
-    SetValueParameterFormat valueParameterName format -> over _valueParameterInfo $ updateValueParameterInfo (setValueParameterFormat format) valueParameterName
-    DeleteValueParameterInfo valueParameterName -> over _valueParameterInfo $ OMap.delete valueParameterName
-    SetChoiceDescription choiceName description -> over _choiceInfo $ updateChoiceInfo (setChoiceDescription description) choiceName
-    SetChoiceFormat choiceName format -> over _choiceInfo $ updateChoiceInfo (setChoiceFormat format) choiceName
-    DeleteChoiceInfo choiceName -> over _choiceInfo $ Map.delete choiceName
-  assign (_hasUnsavedChanges) true
-  where
-  setChoiceDescription :: String -> ChoiceInfo -> ChoiceInfo
-  setChoiceDescription newDescription x = x { choiceDescription = newDescription }
-
-  setChoiceFormat :: NumberFormat -> ChoiceInfo -> ChoiceInfo
-  setChoiceFormat newChoiceFormat x = x { choiceFormat = newChoiceFormat }
-
-  setValueParameterDescription :: String -> ValueParameterInfo -> ValueParameterInfo
-  setValueParameterDescription newDescription x = x { valueParameterDescription = newDescription }
-
-  setValueParameterFormat :: NumberFormat -> ValueParameterInfo -> ValueParameterInfo
-  setValueParameterFormat newValueParameterFormat x = x { valueParameterFormat = newValueParameterFormat }
+carryMetadataAction = do
+  -- | Let's cache generic set actions handlers
+  let
+    genericDispatcher =
+      Record.merge
+        (mkSettersDispatcher (Proxy :: Proxy MetaData))
+        (mkUpdatersDispatcher (Proxy :: Proxy MetaData))
+  \(MetadataAction action) -> do
+    let
+      go =
+        ( (Variant.match genericDispatcher <<< Variant.expand)
```

Comment:
I'm not really sure what is going on here... In the initial code that I saw, I thought all of the handlers were generic, in here I thiiink some are being manually dispatched and some are being genericly dispatched? Some comments to clarify this would be useful. Also, the fact that is not immediately obvious is something to consider 🤔

Reply:
The motivation behind this complexity here is motivated by the large comment in the `MetadataTab/Types.purs` above - I've tried to provide "an example" of composition. I'm not sure if I should leave it as it is or simplify this code altogether (into a fully generic handler) and make example composition of handlers a part of the docs.